dipy/tracking/streamline.py

```diff
@@ -323,3 +327,136 @@ def orient_by_rois(streamlines, roi1, roi2, affine=None, copy=True):
             new_sl[idx] = sl[::-1]
 
     return new_sl
+
+
+def _extract_vals(data, streamlines, affine=None, threedvec=False):
+    """
+    Helper function for use with `values_from_volume`.
+
+    Parameters
+    ----------
+    data : 3D or 4D array
+        Scalar (for 3D) and vector (for 4D) values to be extracted. For 4D
+        data, interpolation will be done on the 3 spatial dimensions in each
+        volume.
+
+    streamlines : ndarray or list
+        If array, of shape (n_streamlines, n_nodes, 3)
+        If list, len(n_streamlines) with (n_nodes, 3) array in
+        each element of the list.
+
+    affine : ndarray, shape (4, 4)
+        Affine transformation from voxels (image coordinates) to streamlines.
+        Default: identity.
+
+    threedvec : bool
+        Whether the last dimension has length 3. This is a special case in
+        which we can use :func:`vfu.interpolate_vector_3d` for the
+        interploation of 4D volumes without looping over the elements of the
+        last dimension.
+
+    Return
+    ------
+    array or list (depending on the input) : values interpolate to each
+        coordinate along the length of each streamline
+    """
+    data = data.astype(np.float)
+    if affine is not None:
+        streamlines = ut.move_streamlines(streamlines,
+                                          np.linalg.inv(affine))
+
+    if (isinstance(streamlines, list) or
+            isinstance(streamlines, types.GeneratorType)):
+        vals = []
+        for sl in streamlines:
+
+            if threedvec:
+                vals.append(list(vfu.interpolate_vector_3d(data,
+                                 sl.astype(np.float))[0]))
+            else:
+                vals.append(list(vfu.interpolate_scalar_3d(data,
+                                 sl.astype(np.float))[0]))
+
+    elif isinstance(streamlines, np.ndarray):
+        sl_shape = streamlines.shape
+        sl_cat = streamlines.reshape(sl_shape[0] *
+                                     sl_shape[1], 3).astype(np.float)
+        if affine is not None:
+            sl_cat = (np.dot(sl_cat, affine[:3, :3]) + affine[:3, 3])
```

Comment:
See the recent commit. Now, the affine is not applied upfront. Instead,
each case is handled separately: if you have a list or generator, it gets
moved by move_streamline. If it's an array, we apply the affine to it here.
To answer your question, this is just a simplified way of applying the
affine. Dot product your coordinates with the top left 3x3 part of the
affine to apply the rotation and then add the last column to apply the
translation (see here for details:
http://nipy.org/nibabel/coordinate_systems.html)

I've also added a test to check that I am not changing the streamlines on
you, but I'd be happy if you could take a look at the data you are putting
through this and see that indeed the coordinates in the streamlines you
pass through remain the same, after passing through this function, in case
I missed some cases.


On Tue, Jan 26, 2016 at 5:55 AM, Samuel St-Jean <notifications@github.com>
wrote:

> In dipy/tracking/streamline.py
> <https://github.com/nipy/dipy/pull/811#discussion_r50836783>:
>
> > +        vals = []
> > +        for sl in streamlines:
> > +
> > +            if threedvec:
> > +                vals.append(list(vfu.interpolate_vector_3d(data,
> > +                                 sl.astype(np.float))[0]))
> > +            else:
> > +                vals.append(list(vfu.interpolate_scalar_3d(data,
> > +                                 sl.astype(np.float))[0]))
> > +
> > +    elif isinstance(streamlines, np.ndarray):
> > +        sl_shape = streamlines.shape
> > +        sl_cat = streamlines.reshape(sl_shape[0] *
> > +                                     sl_shape[1], 3).astype(np.float)
> > +        if affine is not None:
> > +            sl_cat = (np.dot(sl_cat, affine[:3, :3]) + affine[:3, 3])
>
> Oh good, but what does it actually do? It was a real question and not a
> bug report :P I thought if you moved everything to image space with the
> move streamline function, why do you need to reapply the affine?
>
> Might explain why my extracted values seems decent, but whenever I try to
> save back the streamlines they won't fit on anatomy.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nipy/dipy/pull/811/files#r50836783>.
>